reactor-core-micrometer/src/main/java/reactor/core/observability/micrometer/MicrometerObservationListener.java

```diff
@@ -88,47 +88,61 @@
 
 	@Override
 	public void doFirst() {
-		ContextSnapshot contextSnapshot = ContextSnapshot.capture(this.originalContext);
-
-		try (ContextSnapshot.Scope ignored = contextSnapshot.setThreadLocalValues()) {
-			this.scope = this.subscribeToTerminalObservation
-				.start()
-				.openScope();
-			//reacquire the scope from ThreadLocal
-			//tap context hasn't been initialized yet, so addToContext can now use the Scope
-			ContextSnapshot contextSnapshot2 = ContextSnapshot.capture(this.originalContext);
-			this.contextWithScope = contextSnapshot2.updateContext(Context.of(this.originalContext));
+		/* Implementation note on using parentObservation vs openScope:
+		We deliberately don't open nor store Scopes. This is a snake pit, as there's no guarantee that the thread
+		in which doFirst is invoked is going to be the same as the one in which eg. doOnComplete is invoked, in
+		which case we'd get thread pollution (as Scope's role is to put values in ThreadLocals).
+		So this tap listener only deals with Observation and ensuring that Observations are hierarchical including
+		when an inner tap is created in another thread (discovered via Context) or is visible in the current thread
+		(discovered via registry.getCurrentObservation()).
```

Comment:
Nice to have this implementation note! 

That said, it makes me thing, there is a big problem here. Maybe it's just the way it's phased, because it does seem like the code is in a better place now. If so, perhaps it could be re-phrased a little to start with stating what the listener needs to focus on, and then elaborate a bit on why this should not be replaced with openScope instead, and the considerations around the thread model for doFirst and doOnComplete.

Reply:
I've rephrased it a bit in the last commit, does that help?

Reply:
Yes, that looks super helpful !!